### PR TITLE
Lecture 1 setup and product dao

### PR DESCRIPTION
**Pull request by VictoriaZk**

### CODE REVIEW

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 1]
```diff
@@ -1,25 +1,43 @@
...
+        products = new ArrayList<>();
+    }
+
+
     @Override
     public Product getProduct(Long id) {
-        throw new RuntimeException("Not implemented");
+        return products.stream().
+                filter(product -> product.getId().equals(id)).
+                findAny().orElseThrow( () -> new RuntimeException( "Product is not found"));
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: IllegalArgumentException here?

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 2]
```diff
@@ -1,25 +1,43 @@
...
-        throw new RuntimeException("Not implemented");
+        return products.stream().
+                filter(product -> product.getPrice() != null && product.getStock() > 0).
+                collect(Collectors.toList());
     }
 
     @Override
     public void save(Product product) {
-        throw new RuntimeException("Not implemented");
+        if (product == null) {
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: What happens if there is such product in the list already? Should we check this and throw an exception in this case?

Consider rewriting using streams.

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 3]
```diff
@@ -3,19 +3,58 @@
...
     public void testFindProductsNoResults() {
         assertTrue(productDao.findProducts().isEmpty());
     }
+
+    @Test
+    public void testFindProduct() {
+        product.setPrice(new BigDecimal(10));
+        product.setStock(12);
+        productDao.save(product);
+        assertTrue(productDao.findProducts() != null);
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java]: Can be simplified to assertNotNull

File: src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java [hunk 4]
```diff
@@ -3,19 +3,58 @@
...
+    }
+
+    @Test
+    public void testGetProductById() {
+        product.setPrice(new BigDecimal(1));
+        product.setStock(1);
+        productDao.save(product);
+        assertEquals(ID, productDao.getProduct(ID).getId());
+    }
+
```
Comment (SergeiStarikov) [inline src/test/java/com/es/phoneshop/model/product/ArrayListProductDaoTest.java:52]: Consider adding more tests on edge cases: for example, test to add two products with the same id.

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 5]
```diff
@@ -1,43 +1,52 @@
...
     }
 
-
     @Override
-    public Product getProduct(Long id) {
+    public synchronized Product getProduct(Long id) {
         return products.stream().
                 filter(product -> product.getId().equals(id)).
-                findAny().orElseThrow( () -> new RuntimeException( "Product is not found"));
+                findAny().orElseThrow( () -> new ProductNotFoundException( "Product is not found"));
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Here and in other places - formatting:

https://marcin-chwedczuk.github.io/java-streams-best-practices

Compare readability:

```
return products.stream().
                filter(product -> product.getId().equals(id)).
                findAny().orElseThrow(...)
```

VS:

```
return products.stream()
                .filter(product -> product.getId().equals(id))
                .findAny()
                .orElseThrow(...)
```

File: src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java [hunk 6]
```diff
@@ -1,43 +1,52 @@
...
+        try {
+            productWithSameId = getProduct(product.getId());
+        } catch (ProductNotFoundException exception) {
+            productWithSameId = null;
         }
-        else {
+        if (productWithSameId == null) {
             products.add(product);
+        } else {
+            throw new IllegalArgumentException("Product with such id is already exists");
```
Comment (SergeiStarikov) [inline src/main/java/com/es/phoneshop/model/product/ArrayListProductDao.java]: Can we simplify this to:

```
if (products.contains(product)) {
  ...throw exception...
}
if (product!=null) {
  ...add product...
}

```
OR with streams:

```
if (products.stream()
    .anyMatch(...compapre ID here...) {
    ...throw exception...
} else {
    ...add product...
}
```
